keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +337,302 @@ or later. The dual-stack feature requires no changes to this API.
+2. When configured with two cidrs then the values are expected to have different ip families. e.g `--service-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>` or `--service-cluster-ip-range=<ipv6-cidr>,<ipv4-cidr>`. Attempting to configure a cluster with two similar cidrs will generate failures during startup.
+3. It is possible to have a `partially-dual-stack` cluster, with dual-stack pod networking but only a single service cidr. This means that pods in the cluster will be able to send and receive traffic on both ip families, but services which use the `clusterIP` or `ClusterIPs` field will only be able to use the ip family of this flag.
+4. When upgrading a cluster to dualstack, the first element of the `--service-cluster-ip-range` flag MUST NOT change IP families.  As an example, an IPv4 cluster can safely be upgraded to use `--service-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>`, but can not safely be changed to `--service-cluster-ip-range=<ipv6-cidr>,<ipv4-cidr>`.
+5. The first cidr entry of `--service-cluster-ip-range=<first cidr entry>, <secondary cidr entry>` is considered cluster default ip address family.
+6. Attempting to allocate `ClusterIP` or `ClusterIPs` from an ip address family that is not configured on the cluster will result in an error.
+
+> The below discussion assumes that `Endpoint Slice and Endpoint Slice Controller`  will allow dual stack endpoints, the existing endpoint controller will remain single stack. That means enabling dual stack services by having more than one entry in `--service-cluster-ip-range` without enabling `Endpoint Slice` feature gate will result in validation failures during startup.
+
+#### Kubernetes Default Service
+
+> discussion of `spec.preferDualStack` and `spec.ipFamilies` fields is detailed in the next section.
+
+When the api-server is configured with more than one cidr in `--service-cluster-ip-range` flag. api-server will automatically recreate dualstack enabled kubernetes default service. The service will carry `.spec.ClusterIP` of the same ip family as cluster default ip address family. The rest of the fields will be set according to the same service below for a cluster configured with `--service-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>`
+
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: kuberentes
+spec:
+  type: ClusterIP
+  preferDualStack: true
+  ipFamilies: # the order of the families is according to how they are configured on the cluster
+    - IPv4 # cluster default
+    - IPv6 
+  clusterIP: <ipv4> # allocated
+  clusterIPs:
+    - <ipv4>
+    - <ipv6> 
+   <snip>
+```
+
+Kubernetes default service is expressed current in pods as environment variables, these variables offers location, ip and other details of kubernetes default service. These variables will remain. The following variables will be added
+
+```
+# for IPv4
+KUBERNETES_IPv4_PORT=tcp://<IPv4>:443
```

Comment:
Are we going to give all service env vars this treatment?  I don't think we should.  Do we really need to do this even for KUBERNETES ?

Reply:
I _really_ didn't want to mess with env vars. I am kind of hoping you ask me to remove :-) this part.

Reply:
No pod ever _needs_ to know the apiserver's secondary service IP. In any cluster, whether created dual-stack or upgraded from single-stack, all apiservers will always be listening on the primary service IP family, and all pods will always have outbound connectivity on the primary service IP family. So the existing environment vars are sufficient to guarantee that all pods can reach the apiserver.

Reply:
Could NetworkPolicy be an issue here? Eg an originally single-stack IPv6 cluster with a Pod that is only permitted to sent IPv4 traffic.

Reply:
I suppose in that case the pod would not be able to use `rest.InClusterConfig()`. If that's a problem then "don't do that then"?

Note that the pod would still be able to reach `kubernetes.default.svc.cluster.local`.

Reply:
I think we should document that service envvars only cover `clusterIPs[0]`, remove these changes, and see if anyone screams during the next cycle(s).  We can always come back to it, right?

Reply:
yes. that can be added later. Thanks. will remove.